### PR TITLE
telemetry: report heartbeat metrics as JSON numbers

### PR DESCRIPTION
**Pull request by cce**
## Summary

The current "Heartbeat" telemetry event uses JSON strings inside a `Metrics` object to encode counter values. This PR sends them as JSON numbers instead in a new `m` object. For backwards compatibility the StringGauge `version`, `version-num`, `channel`, `branch`, and `commit-hash` string metadata included in the old `Metrics` object are preserved.

This PR also fixes a bug where counters with multiple labels were not being reported correctly (only one of the labeled values was being reported, with each iteration over the labels overwriting the reported value for the the base metric name). Label names and values are now added to the reported metric names, sanitized with `_` for non-alphanumeric unsupported characters.

## Test Plan

Slightly updated old tests. Added a new test to assert the new message JSON serialization created by calling `logTelemetry` with this event type.

### CODE REVIEW

**Review by winder — DISMISSED** [review]
Seems fine to me. You force telemetry to start locally with `-t 1`, that would create the index for devnet so you can verify that the mapping is correct.

File: logging/telemetryspec/event.go [hunk 1]
```diff
@@ -43,7 +43,14 @@ const HeartbeatEvent Event = "Heartbeat"
...
 // HeartbeatEventDetails contains details for the StartupEvent
 type HeartbeatEventDetails struct {
-	Metrics map[string]string
+	Info struct {
+		Version    string `json:"version"`
+		VersionNum string `json:"version-num"`
+		Channel    string `json:"channel"`
+		Branch     string `json:"branch"`
+		CommitHash string `json:"commit-hash"`
+	} `json:"Metrics"` // backwards compatible name
```
Comment (winder) [inline logging/telemetryspec/event.go:52]: Nice :+1:

File: logging/telemetry_test.go [hunk 2]
```diff
@@ -213,6 +214,44 @@ func TestDetails(t *testing.T) {
...
+	}
+	f.telem.logEvent(f.l, telemetryspec.ApplicationState, telemetryspec.HeartbeatEvent, hb)
+
+	data := f.hookData()
+	a.NotNil(data)
+	a.Len(data, 1)
+	a.Equal(hb, data[0]["details"])
+
+	// assert JSON serialization is backwards compatible
+	js, err := json.Marshal(data[0])
```
Comment (algorandskiy) [inline logging/telemetry_test.go:239]: maybe just store an old serialized value and compare against new?

**Review by algorandskiy — COMMENTED** [review]
Looks good, one question

File: logging/telemetry_test.go (hunk 2 above)
Comment (cce) [reply logging/telemetry_test.go:239]: I couldn't figure out how to write a test that had access to the serialized value that the telemetry system makes ... there is a JSON conversion going on inside the logrus hook in a third party library, before being sent via the telemetry client. So I figured I would just simulate that call to json.Marshal here..

File: util/metrics/registryCommon.go [hunk 3]
```diff
@@ -25,11 +26,17 @@ import (
...
 
 // Registry represents a single set of metrics registry
 type Registry struct {
 	metrics   []Metric
 	metricsMu deadlock.Mutex
 }
+
+var sanitizeTelemetryCharactersRegexp = regexp.MustCompile("(^[^a-zA-Z_]|[^a-zA-Z_0-9])")
+
+func sanitizeTelemetryName(name string) string {
```
Comment (brianolson) [inline util/metrics/registryCommon.go:40]: I'd love to see a test for this function. It should be easy to write and would demonstrate what that slightly messy regex is supposed to achieve. And a comment, like, "ensure telemetry name starts with `[a-zA-Z0-9_]`"

**Review by brianolson — COMMENTED** [review]
Lots of good. I think one more bit of comment+test will make it more durable.

File: util/metrics/registryCommon.go (hunk 3 above)
Comment (cce) [reply util/metrics/registryCommon.go:43]: thanks, I have added a test and comment, and also noticed another test I was modifying was build-tagged out of being run by CI, updated it too.

**Comment by cce** [comment]
@winder @brianolson @algorandskiy addressed CR feedback and ready for re-review

**Review by brianolson — APPROVED** [review]

**Review by winder — APPROVED** [review]

**Pull request merged**